fontforge/fvfonts.c

```diff
@@ -311,7 +311,7 @@ static AnchorPoint *AnchorPointsDuplicate(AnchorPoint *base,SplineChar *sc) {
     AnchorPoint *head=NULL, *last=NULL, *cur;
     AnchorClass *ac;
 
-    for ( ; base!=NULL; base = base->next ) {
+    for ( ; base!=NULL && sc->parent!=NULL; base = base->next ) {
```

Comment:
I haven't looked at the conditions that lead to this problem but just looking at the loop itself I think I would be more comfortable with

```
    for ( ; base!=NULL; base = base->next ) {
        if (sc->parent==NULL)
            continue;
```
That way you still cover all the `base`s (as it were).

Reply:
What would that solve? The entire loop can be skipped over because `sc` and `sc->parent` are never changed within the body of the loop. It's pointless to iterate over and do that check a ton of times when we know that if it's NULL at the start, it will always be NULL the duration of the function.

Reply:
Actually, on that note, I suppose I could just change it to:

```c
    if (sc->parent == NULL)
        return head;
    for ( ; base!=NULL; base = base->next ) {
```

But having the check be part of the loop body seemed more elegant to me, because then there's still only one `return` statement at the bottom of the function.

Walking through a linked list and explicitly not doing anything with any of the elements seems a bit pointless.. Though I suppose with my proposed change, it does *both* checks for every element (when the function is called correctly without `sc->parent` being NULL), when it doesn't actually need to. That's a point in favor of having a check before the loop that adds another `return` statement.

Reply:
Ah, I suppose I *could* just wrap the whole `for` loop in an `if` statement that checks if it's *not* NULL. That would both ensure the check only happens once, *and* keeps only a single `return` statement.

However, that would break Linux kernel guidelines, which state to not indent more than 3 times.. I've found that rule to be bizarrely very good at fixing bugs for no discernible reason, so I try to stick with it. The function already breaks it with 4-levels of indentation at times, but this would be bumping that up to 5 levels of indentation. Besides, I've heard that it's better to avoid wrapping things in `if` statements like that if you can just have a small `if` statement that returns early instead.

I could force things by instead of an extra `return` statement, have a `goto` that jumps to the return statement at the bottom.. But `goto` is discouraged for good reason and I feel like that's a rather silly use for one.

What are your thoughts, @skef?

Reply:
... Actually, come to think of it, maybe this is the wrong place to change *any*thing. Maybe it'd be best to just.. Remove the function call. That is, make sure that the function isn't called if `sc->parent` is NULL to begin with. The function is only called in a single place (within `SplineCharCopy()`), so we could just wrap the function call in an `if` statement.

That would be more elegant anyway, because that's *also* where `sc->parent` gets set to NULL to begin with, currently on line 531 of fontforge/fvfonts.c:

```c
    nsc->parent = into;
```

We could just change line 542 (the call to `AnchorPointsDuplicate()`) to be:

```c
    if (into)
        nsc->anchor = AnchorPointsDuplicate(nsc->anchor,nsc);
```

And that should be just as good. After all, if we're detecting that we're copying things into NULL, we should just.. Not do that.

Reply:
Turns out there's another line in that function that gives special handling for when `into` is set to `NULL` or `0`: line 505:

```c
    nsc->layer_cnt = into==NULL?2:into->layer_cnt;
```

Indeed, if I try to *just* wrap that `nsc->anchor = ...` line in an `if` statement, trying to open any file causes a crash. It has to be set to `NULL` instead of being left as-is.

The fix that works is changing line 542 to this:

```c
    nsc->anchor = into==NULL?NULL:AnchorPointsDuplicate(nsc->anchor,nsc);
```

Reply:
I reviewed this in a hurry and didn't think my advice entirely through. I think it's fine to return NULL before the loop if sc->parent is NULL. That makes it clear why the flow doesn't progress through the loop.